### PR TITLE
CUMULUS-1449 Terraform Config Updates

### PR DESCRIPTION
**Pull request by nemreid**
**Summary:** Summary of changes

Updates Terraform Configuration for Cumulus 1449 and adds additional compatibility fixes (e.g. CUMULUS-1394 compatibility.)

### CODE REVIEW

File: example/cumulus-tf/cnm_response_task.tf [hunk 1]
```diff
@@ -1,7 +1,7 @@
 resource "aws_lambda_function" "cnm_response_task" {
```
Comment (nemreid) [inline example/cumulus-tf/cnm_response_task.tf]: These functions will additionally need to be updated with the CMA lambda layer, and probably require some more fixes before `KinesisTestTriggerSpec` will pass. cc @Jkovarik

File: example/cumulus-tf/cnm_to_cma_task.tf [hunk 2]
```diff
@@ -1,7 +1,7 @@
 resource "aws_lambda_function" "cnm_to_cma_task" {
```
Comment (nemreid) [inline example/cumulus-tf/cnm_to_cma_task.tf]: These functions will additionally need to be updated with the CMA lambda layer, and probably require some more fixes before KinesisTestTriggerSpec will pass. cc @Jkovarik

File: tf-modules/ingest/reporter-functions.tf [hunk 3]
```diff
@@ -0,0 +1,32 @@
+module "report_executions" {
```
Comment (nemreid) [inline tf-modules/ingest/reporter-functions.tf]: Ideally these modules would be part of the archive module, because they are really only thin lambda wrappers around DynamoDB PUT operations, but I did not have the time to resolve the dependencies that would bring today.

File: example/app/config.yml [hunk 4]
```diff
@@ -61,7 +61,7 @@ default:
   # test configuration ends
 
   cmaDir: '/opt/'
-  sandbox_cma_layer: arn:aws:lambda:us-east-1:{{AWS_ACCOUNT_ID}}:layer:CMA_TEST:47
+  sandbox_cma_layer: arn:aws:lambda:us-east-1:{{AWS_ACCOUNT_ID}}:layer:Cumulus_Message_Adapter:9
```
Comment (Jkovarik) [inline example/app/config.yml]: 👍

File: packages/common/util.js [hunk 5]
```diff
@@ -192,3 +193,15 @@ exports.lookupMimeType = (key) => mime.lookup(key) || null;
...
+ * Pass a value through a pipeline of functions and return the result
+ *
+ * @param {*} value - the value to be passed through the pipeline of functions
+ * @param {...Function} fns - the functions to be invoked
+ * @returns {*} the result of passing the value through the functions
+ *
+ * - If no functions are provided, the value is returned.
+ * - Functions should expect a single argument
+ */
+exports.thread = (value, ...fns) => flow(fns)(value);
```
Comment (Jkovarik) [inline packages/common/util.js]: Why not just bring in flow directly where needed?  I can see it for dependency management/convenience, but in that case,  the name `thread` doesn't feel great.   I think we should use a name  that references flow, or at least doesn't bring to mind the other context 'threads/threading' can be used in.

File: packages/common/string.js [hunk 6]
```diff
@@ -52,6 +52,22 @@ function globalReplace(string, oldSubString, newSubString) {
...
+ * @param {string} newSubstr - The string replacement
+ * @param {string} string - The string to modify
+ * @returns {string} the modified string
+ *
+ * This is a curried function - https://lodash.com/docs/4.17.11#curry
+ *
+ * @static
+ */
+const globalReplaceC = curry(
+  (substr, newSubstr, string) => globalReplace(string, substr, newSubstr)
```
Comment (Jkovarik) [inline packages/common/string.js]: As annoying as more typing is, I'm not immediately on-board with the "C" as "curried" function convention.     Consider `globalReplaceCurried` for readability instead. 

Also consider deprecating `globalReplace`, or just wholesale updating it.   We can always publish it as a breaking change in the CL, *particularly* given TF is going to be a major release.

File: example/spec/helpers/granuleUtils.js [hunk 7]
```diff
@@ -88,45 +90,57 @@ function createGranuleFiles(granuleFiles, bucket, oldGranuleId, newGranuleId) {
...
     bucket,
     oldGranuleId,
     newGranuleId
   );
 
-  const baseInputPayloadJson = JSON.stringify(baseInputPayload);
-  const updatedInputPayloadJson = globalReplace(baseInputPayloadJson, oldGranuleId, newGranuleId);
-
-  return JSON.parse(updatedInputPayloadJson);
+  return thread(
```
Comment (Jkovarik) [inline example/spec/helpers/granuleUtils.js]: I like this use of the curried globalReplace function. 👍

File: example/spec/helpers/granuleUtils.js [hunk 8]
```diff
@@ -88,45 +90,57 @@ function createGranuleFiles(granuleFiles, bucket, oldGranuleId, newGranuleId) {
...
-  return JSON.parse(fileContentsWithIdPathAndCollection);
-}
+const loadFileWithUpdatedGranuleIdPathAndCollection = (
+  filename,
+  newGranuleId,
+  newPath,
+  newCollectionId,
+  stackId
+) => thread(
+  fs.readFileSync(filename, 'utf8'),
```
Comment (Jkovarik) [inline example/spec/helpers/granuleUtils.js]: 👍

File: example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js [hunk 9]
```diff
@@ -348,15 +347,9 @@ describe('When there are granule differences and granule reconciliation is run',
     // need to add the cmr granule back to the table, so the granule can be removed from api
     await granuleModel.create(cmrGranule);
     await granulesApiTestUtils.removeFromCMR({ prefix: config.stackName, granuleId: cmrGranule.granuleId });
-    await waitForConceptExistsOutcome(cmrGranule.cmrLink, false);
```
Comment (Jkovarik) [inline example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js]: I'm sure it should be obvious, but why was this removed?

File: example/spec/helpers/granuleUtils.js [hunk 10]
```diff
@@ -88,45 +90,57 @@ function createGranuleFiles(granuleFiles, bucket, oldGranuleId, newGranuleId) {
  * @returns {Promise<Object>} - input payload as a JS object with the updated granule ids
  */
 async function setupTestGranuleForIngest(bucket, inputPayloadJson, granuleRegex, testSuffix = '', testDataFolder = null) {
-  // granule id for the new files
-  const newGranuleId = randomStringFromRegex(granuleRegex);
-  console.log(`\ngranule id: ${newGranuleId}`);
+  const payloadJson = testDataFolder ?
```
[the remark this answers is not in the document]
Comment (yjpa7145) [reply example/spec/helpers/granuleUtils.js]: This change wasn't done with terseness in mind, it was to prevent reassigning a function parameter. It could be refactored to this if you would prefer:

```js
let payloadJson;
if (testDataFolder) {
  payloadJson = globalReplaceC('cumulus-test-data/pdrs', testDataFolder, inputPayloadJson);
} else {
  payloadJson = inputPayloadJson;
}
```

File: example/spec/helpers/granuleUtils.js (hunk 7 above)
Comment (yjpa7145) [reply example/spec/helpers/granuleUtils.js]: If you like that, just _wait_ until you see line 137 😄

File: example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js (hunk 9 above)
Comment (yjpa7145) [reply example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js]: That's a question for @nemreid , I'm not sure.

File: example/spec/helpers/granuleUtils.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (Jkovarik) [reply example/spec/helpers/granuleUtils.js]: Sorry, I had the above a bit off.  Edited.   In either case, more for your consideration than a legit PR stopping concern.

File: example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js (hunk 9 above)
Comment (Jkovarik) [reply example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js]: Looks like it was pulled here https://github.com/nasa/cumulus/commit/c679e499955c7fc10c24e0a79f425042221ff2ab#diff-75c457041fb29817c2af29ba14ad2b4e

Comment (Jkovarik) [reply example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js]: Is this just a matter of master not being pulled into the TF branch?

File: tf-modules/ingest/reporter-functions.tf (hunk 3 above)
Comment (yjpa7145) [reply tf-modules/ingest/reporter-functions.tf]: Done.
